src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs

```diff
@@ -421,47 +523,120 @@ internal static void Unescape(ReadOnlySpan<byte> source, Span<byte> destination,
 
 #if BUILDING_INBOX_LIBRARY
                         var rune = new Rune(scalar);
-                        int bytesWritten = rune.EncodeToUtf8(destination.Slice(written));
+                        bool success = rune.TryEncodeToUtf8(destination.Slice(written), out int bytesWritten);
 #else
-                        EncodeToUtf8Bytes((uint)scalar, destination.Slice(written), out int bytesWritten);
+                        bool success = TryEncodeToUtf8Bytes((uint)scalar, destination.Slice(written), out int bytesWritten);
 #endif
+                        if (!success)
+                        {
+                            goto DestinationTooShort;
+                        }
+
                         Debug.Assert(bytesWritten <= 4);
                         written += bytesWritten;
-                    }
+                        break;
                 }
-                else
+
+                if (++idx == source.Length)
                 {
-                    destination[written++] = currentByte;
+                    goto Success;
+                }
+
+                if (source[idx] != JsonConstants.BackSlash)
```

Comment:
Please write tests which hit this new code

Reply:
This is not adding new code as much as it is refactoring how we handle non-escaped characters following escaped characters. As such it is covered both by our existing test coverage in string escaping (which is fairly comprehensive) _and_ new tests added specifically for the newly exposed methods (they both call into the same unescaping routine).